keps/sig-node/20200310-ensure-secret-pulled-images.md

```diff
@@ -0,0 +1,169 @@
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+
+### Goals
+
+Add a flag processed by `kubelet` for `ensureSecretPulledImages` (or something
+similarly named) that, if true, would force `kubelet` to attempt to pull every
+image that was pulled with image pulled secret based authentication, regardless
+of the container image's pull policy.
+
+Optimize to only force re-authentication for a pod when the secret used to pull
+the container image is not present.
+
+### Non-Goals
+
+Out of scope for this KEP is an image caching policy that would direct container
+runtimes through the CRI wrt. how they should treat the caching of images on a
+node. Such as store for public use but only if encrypted. Or Store for private
+use unencrypted...
+
+## Proposal
+
+When `ensureSecretPulledImages` is set, `kublet` will check keep a list of
+container images that required authentication. `kublet` will ensure any image
+in the list is always pulled thus enforcing authentication / re-authentication
+with the exception of pods with secrets containing an auth that has been
+authenticated.
+
+### User Stories
+wip
+
+### Risks and Mitigations
+
+With the default being false, devops engineers may not know to set the flag to
+true in new/old pod specs that are using secrets for pull authentication with
+registries.
+
+A mitigation would be an admission plugin to inject `ensureSecretPulledImages.`
+
+Images authentications with a registry may expire. To mitigate expirations a
+a timeout could be used to force re-authentication.
+
+## Design Details
+
+### Test Plan
+
+tbd
+
+### Graduation Criteria
+
+tbd
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+##### Alpha -> Beta Graduation
+
+tbd
+
+##### Beta -> GA Graduation
+
+tbd
+
+## Implementation History
+
+tbd
+
+## Drawbacks [optional]
+
+Why should this KEP _not_ be implemented. N/A
+
+## Alternatives [optional]
+
+Default the ensure secrets rule to true and don't introduce a new pod spec flag.
+Instead of a pod spec flag make the option a kublet configuration switch or
+set the flag at some other scope.
+
```

Comment:
Perhaps slightly off topic but it feels like another solution to this problem (which doesn't involve Kubelet changes... at least directly) is for the image repository to be high availability? Not sure how that relates to this KEP, but just want to toss it out as another way to deal with pulling images being in the critical path.

Reply:
Different architectures for ensuring container images on nodes for pods can be contemplated, but that still leaves the does pod b have rights to pod a's authenticated images problem. IOW it's not just about HA it's also about security/privacy. For example, a multi-tenant environment user A may have secure data in a blob in a container that can only be pulled from a private registry with authentication.  User b may not want that image to be available to other users/pods.

Reply:
I think I'm a little confused about this - if we are pulling every time from a HA image repo, how does pod B have rights to pod A's authenticated images? Doesn't pod B have to be able to pull the image (and thus be auth'd) in order to use it?

Reply:
The only way to "force" pod B to pull with auth is for pod B to either edit the pod spec manually or to run the pull always injector that inserts the pull always flag into the pod spec irregardless. This is not desirable unless you want to always pull the manifest every single time for every single pod for every single image used (sans the pause container). This kep is to ensure secret pulled image authentications without using the current recommend technique of injecting pull always into every pod spec. For short lived containers/pods pull always is not an ideal situation performance wise and for all containers/pods this solution does not work if the registry is unavailable.

Reply:
Registries are generally remote from the host/worker node and employ their own HA techniques. Container runtimes keep a cache of images on the node and do the pull requests to these remote registries for kublet through the CRI. Kubelet could load the images into a container runtime's cache avoiding CRI pull request altogether. So yes we could come up with a different architecture where k8s has an ha container image server in the cluster for use by the nodes to load images just in time for creating pods. Simply a matter of code and configuration and migration and... :)